syntax = "proto3";

package modelservice;

service ModelService {
  rpc GenerateContent (GenerateContentRequest) returns (GenerateContentResponse);
  rpc UpdateToken (UpdateTokenRequest) returns (UpdateTokenResponse);
  rpc VerifyTokens (VerifyTokensRequest) returns (VerifyTokensResponse);
  rpc PrepareSpeculative (PrepareSpeculativeRequest) returns (PrepareSpeculativeResponse);
}

message FloatArray {
  repeated float values = 1;
}

message FloatArray2D {
  repeated FloatArray rows = 1;
}

message UpdateTokenRequest {
  string user_uuid  = 1;
  int64 index = 2;
  string input_text = 3;
}

message UpdateTokenResponse {
  bool updated = 1;
}

message VerifyTokensRequest {
  string user_uuid = 1;
  string token_text = 2;
}

message VerifyTokensResponse {
  bool verified = 1;
  bool finished = 2;
  int64 passed_tokens = 3;
  string verified_tokens = 4;
}

message PrepareSpeculativeRequest {
  string user_uuid = 1;
  string prompt = 2;
  int64 max_length = 3;
  int64 generate_step = 4;
  bool exact_mode = 5;
  bool debug_mode = 6;
}

message PrepareSpeculativeResponse {
  string first_tokens = 1;
}

message GenerateContentRequest {
  string user_uuid = 1;
  string prompt = 2;  // 输入提示，可以为空字符串
  int64 max_length = 3;
}

message GenerateContentResponse {
  string generated_text = 1;  // 服务器生成的文本
}
